toolchain/driver/clang_runner.cpp

```diff
@@ -42,24 +51,214 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+  // the runtimes and subsequently the cached build can be used.
+  CARBON_VLOG("Building target resource dir...\n");
+  auto tmp_result = Filesystem::MakeTmpDir();
+  if (!tmp_result.ok()) {
+    CARBON_VLOG("Error making a temporary directory: {0}", tmp_result.error());
+    return false;
+  }
+  Filesystem::RemovingDir tmp_dir = *std::move(tmp_result);
+
+  // Hard code the subdirectory for the resource-dir runtimes.
+  //
+  // TODO: This should be replaced with an abstraction that manages the layout
+  // of a built runtimes tree.
+  std::filesystem::path resource_dir_path =
+      tmp_dir.abs_path() / "clang_resource_dir";
+
+  auto result =
+      BuildTargetResourceDir(target, resource_dir_path, tmp_dir.abs_path());
+  if (!result.ok()) {
+    // FIXME: Need to diagnose this.
+    return false;
+  }
+
+  return RunInternal(args, target, resource_dir_path.native());
+}
+
+auto ClangRunner::RunTargetIndependentCommand(
+    llvm::ArrayRef<llvm::StringRef> args) -> bool {
+  std::string target = ComputeClangTarget(args);
+  return RunInternal(args, target, std::nullopt);
+}
+
+auto ClangRunner::BuildTargetResourceDir(
+    llvm::StringRef target, const std::filesystem::path& resource_dir_path,
+    const std::filesystem::path& tmp_path) -> ErrorOr<Success> {
+  // Disable any leaking of memory while building the target resource dir, and
+  // restore the previous setting at the end.
+  auto restore_leak_flag = llvm::make_scope_exit(
+      [&, orig_flag = enable_leaking_] { enable_leaking_ = orig_flag; });
+  enable_leaking_ = false;
+
+  // Create the destination directory if needed.
+  CARBON_ASSIGN_OR_RETURN(
+      Filesystem::Dir resource_dir,
+      Filesystem::Cwd().CreateDirectories(resource_dir_path));
+
+  // Symlink the installation's `include` and `share` directories.
+  std::filesystem::path install_resource_path =
+      installation_->clang_resource_path();
+  CARBON_RETURN_IF_ERROR(
+      resource_dir.Symlink("include", install_resource_path / "include"));
+  CARBON_RETURN_IF_ERROR(
+      resource_dir.Symlink("share", install_resource_path / "share"));
+
+  // Create the target's `lib` directory.
+  CARBON_ASSIGN_OR_RETURN(
+      Filesystem::Dir lib_dir,
+      resource_dir.CreateDirectories(std::filesystem::path("lib") /
+                                     std::string_view(target)));
+
+  llvm::Triple target_triple(target);
+  CARBON_CHECK(!target_triple.isOSWindows(),
+               "TODO: Windows runtimes are untested and not yet supported.");
+
+  // Create a directory with a known path that we can use for object files. We
+  // use an explicit `_objs` subdirectory for clarity when debugging.
+  CARBON_ASSIGN_OR_RETURN(Filesystem::Dir tmp_dir,
+                          Filesystem::Cwd().OpenDir(tmp_path));
+  CARBON_ASSIGN_OR_RETURN(Filesystem::Dir objs_dir,
+                          tmp_dir.OpenDir("_objs", Filesystem::OpenAlways));
+  std::filesystem::path objs_path = tmp_path / "_objs";
```

Comment:
Do we need to create the CRT files in the "wrong" place and then moving them to the intended location? Clang already does that dance itself (changing the filename not the directory), so I think this will result in Clang creating the file in the wrong directory with the wrong filename, then Clang renaming it from the wrong filename to the right filename in the wrong directory, then us renaming it from the wrong directory to the right directory.

If this is important, using the same name `_objs` that bazel uses seems potentially confusing; maybe something like `build` (or just dropping the leading `_`) would be clearer that this isn't being built by bazel.

Reply:
It's important for the builtins archive -- we don't ever want the objects there.

I'd done the rename here mostly for consistency with the archive, and to consistently not depend on having an absolute path for the resource directory.

I can switch back if you prefer, but it will mean that we *must* be able to spell the resource directory path as part of the Clang argument list, rather than only requiring that of the temporary path. I guess we can cross that bridge if we ever end up caring.

I've just removed the `_objs` path component and passed the temp directory on through. Not sure the extra directory component was adding any value here.